include/rmm/mr/device/failure_alternate_resource_adaptor.hpp

```diff
@@ -0,0 +1,184 @@
+  failure_alternate_resource_adaptor(failure_alternate_resource_adaptor&&) noexcept =
+    default;  ///< @default_move_constructor
+  failure_alternate_resource_adaptor& operator=(failure_alternate_resource_adaptor&&) noexcept =
+    default;  ///< @default_move_assignment{failure_alternate_resource_adaptor}
+
+  /**
+   * @briefreturn{rmm::device_async_resource_ref to the upstream resource}
+   */
+  [[nodiscard]] rmm::device_async_resource_ref get_upstream_resource() const noexcept
+  {
+    return primary_upstream_;
+  }
+
+  /**
+   * @briefreturn{rmm::device_async_resource_ref to the alternate upstream resource}
+   */
+  [[nodiscard]] rmm::device_async_resource_ref get_alternate_upstream_resource() const noexcept
+  {
+    return alternate_upstream_;
+  }
+
+  /**
+   * @briefreturn{PrimaryUpstream* to the upstream memory resource}
+   */
+  [[nodiscard]] PrimaryUpstream* get_upstream() const noexcept { return primary_upstream_; }
+
+  /**
+   * @briefreturn{AlternateUpstream* to the alternate upstream memory resource}
+   */
+  [[nodiscard]] AlternateUpstream* get_alternate_upstream() const noexcept
+  {
+    return alternate_upstream_;
+  }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource.
+   *
+   * @throws `exception_type` if the requested allocation could not be fulfilled
+   * by the primary or the alternate upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    void* ret{};
+    try {
+      ret = primary_upstream_->allocate(bytes, stream);
+    } catch (exception_type const& e) {
+      ret = alternate_upstream_->allocate(bytes, stream);
+      std::lock_guard<std::mutex> lock(mtx_);
+      alternate_allocations_.insert(ret);
+    }
+    return ret;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `ptr`
+   *
+   * @param ptr Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* ptr, std::size_t bytes, cuda_stream_view stream) override
+  {
+    std::size_t count{0};
+    {
+      std::lock_guard<std::mutex> lock(mtx_);
+      count = alternate_allocations_.erase(ptr);
+    }
+    if (count > 0) {
+      alternate_upstream_->deallocate(ptr, bytes, stream);
+    } else {
+      primary_upstream_->deallocate(ptr, bytes, stream);
+    }
+  }
+
+  /**
+   * @brief Compare the resource to another.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  [[nodiscard]] bool do_is_equal(device_memory_resource const& other) const noexcept override
```

Comment:
Not sure how we handle `do_is_equal` once the template parameters have gone? @harrism?

Reply:
I used the `==` operator

Reply:
This code is how I've been writing the other adaptor comparisons.